### DISCUSSION TITLE
A concerning trend in export size

### DISCUSSION THREAD

**Discussion by R2bEEaton** [Budibase/budibase#6840] [Help] [answered]
Hello there!

Over the past month of using Budibase for developing an app, I have created semi-regular exports just to save progress in case I break something. I've noticed a concerning trend: the size of the exports is increasing more than expected. Below are a couple of simple graphs I made in Google Sheets to illustrate this point.

![image](https://user-images.githubusercontent.com/34921506/180206658-37a7a60f-a893-4704-ae83-f02c56b7a2f9.png)

Obviously it would be expected that as my app's complexity increased, the export size would as well, but some of these increases don't make much sense to me. The most glaring one is that *today* after updating from `v1.0.218` to `v1.1.17`, my file size increased by ~24MB. (Edit: Ahhh! Six hours later it's now 91MB!) The app hasn't changed much at all since July 11th, yet the export size has almost doubled. Where's it all coming from? I only have ~250 internal rows saved, ~8 simple screens, and ~2 more complex screens.

The graph on the right shows the number of "_deleted" rows in each export. I noticed in #5075 that someone was having a problem with an excess of deleted rows. During testing, I've added and removed many rows, which I don't suspect are being cleared out as mentioned in the linked discussion. Perhaps this is the source of some of the file size increase, but probably not all.

How much impact does this app export size have on performance anyway? This could be entirely unrelated, but I've noticed high CPU usage followed by a "Bad Gateway" when navigating to the website - almost as if it crashed - then it comes back as normal. I get 25% CPU usage spikes when refreshing a single page.

### Self-hosting information ###
I'm using the Docker install on a CentOS 7 Virtual Machine running in Windows Hyper-V (if this not ideal, let me know!).
The VM has been given 4 CPU cores (because of occasionally high CPU usage) and 8 GB of RAM (because of high RAM usage, though I have been told this has been fixed in recent versions which I have just updated to).

**Comment by mike12345567** [comment, marked as answer]
Hi @R2bEEaton - thanks for your in depth analysis on this!

Overall, the size of the export won't affect the performance of the system, Couch is able to handle a very large amount of documents, but the point you raise about exported images including the `_deleted` documents is valid, we also need to look into a way to make sure that self hosted installations are running cleanups on their deleted documents regularly, but that is a separate issue.

I've raised [this PR](https://github.com/Budibase/budibase/pull/6867) to remove deleted documents when exporting as they serve no value during an export, in the future there will be some value in keeping the deleted documents for a short while in the case of undo functionality - but this is still in the pipeline.

I expect the issue you were seeing with occasional CPU spikes is due to the instance crashing as there is increased load at startup, I expect this is due to the memory leaks that were seen in previous versions, these should be resolved and we will be carrying out a larger piece of work to remove the library which has been causing our memory leaks, to put the issue to bed once and for all.

**Reply by R2bEEaton** [reply]
Thanks for that PR, looks great! And thanks for the info about document amounts.

But as for the apparent crashing, I don't think it's due to increased load at startup. I've given the container ample resources and have been monitoring it recently. It idles at low usage, but sometimes simply having multiple clients refresh a particularly demanding page of my app can lead to a spike in CPU usage then a 500 "Gateway Error."

I can try to collect more info next week on what causes this for my specific use case. Thanks!

**Reply by R2bEEaton** [reply]
Here's an image of the system load for the past week. It seems one of the recent update has been helpful for the memory leak and other problems. If the apparent crashing happens again, I will let you know.

It should also be noted that this past week has seen decreased usage of our internal app, which may also be a compounding factor.

![image](https://user-images.githubusercontent.com/34921506/182165921-91f5344b-941b-4e61-86f0-51ca490a3089.png)

**Reply by mike12345567** [reply]
Hi @R2bEEaton - thank you for this, good to see proof of a decreased load, very interesting to see! We still have plans for the period of August/September to do a larger rework that should remove the offending library and decrease usage again, so hopefully this will continue to be a trend!
